net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,395 @@
+	}
+
+	err = fifo_set_limit(q->prio_array[prio], 
+			    (q->prio_array[prio])->limit - 1);
+	if (err)
+		return err;
+	
+
+	err = fifo_set_limit(your_queue, your_queue->limit + 1);
+	if (err) 
+		return err;
+	
+	ret = qdisc_enqueue(skb, your_queue, to_free);
+	if (ret != NET_XMIT_SUCCESS) {
+		if (net_xmit_drop_count(ret))
+			qdisc_qstats_drop(sch);
+		return ret;
+	}
+
+	qdisc_qstats_backlog_inc(sch, skb);
+	sch->q.qlen++;
+	return NET_XMIT_SUCCESS;
+
+        }
+
+	if (your_queue->limit == 0) {
+		if (your_prio == 1 ||  
+		   (q->prio_array[your_prio + 1])->limit == 0) 
+			goto done;
+	}
+
+	/* Required priority queue is completey borrowed from and is not able 
+	 * to borrow either. No choice but to drop packet now.
+	 */
+
+	ret = qdisc_enqueue(skb, your_queue, to_free);
+	if (ret != NET_XMIT_SUCCESS) {
+		if (net_xmit_drop_count(ret))
+			qdisc_qstats_drop(sch);
+		return ret;
+	}
+	
+	qdisc_qstats_backlog_inc(sch, skb);
+	sch->q.qlen++;
+	return NET_XMIT_SUCCESS;	
+
+	}
+
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
+done:
+	q->last_dropped_prio = your_prio;
+	ret = qdisc_drop(skb, sch, to_free);
+	return ret;
+
+}
+
+static struct sk_buff *gkprioq_dequeue(struct Qdisc *sch)
+{
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	int err;
+	
+
+	skb = qdisc_dequeue_head(q->prio_array[q->highest_prio]);
+
+	if (skb) {
+		if (q->last_dropped_prio != -1) {
+			err = fifo_set_limit(q->prio_array[q->highest_prio], 
+				   (q->prio_array[q->highest_prio])->limit - 1);
+			if (err)
+				pr_warn_once("Something is wrong!");
+			
+			/* Donate resources to priority queue whose last packet 			 * was dropped because of resource unavailability then.
+	                 */
+
+			err = fifo_set_limit(q->prio_array[q->last_dropped_prio]			     ,(q->prio_array[q->last_dropped_prio])->limit + 1);
+
+			if (err)
+				pr_warn_once("Something is wrong!");
+	
+		}
+
+		goto done; 
+	}
+
+	return NULL;
+
+done:
+	if ((q->prio_array[q->highest_prio])->q.qlen > 0)
+		return skb;
+
+	q->highest_prio = calc_new_high_prio(q); 
+	return skb;
+
+}
+
+static int gkprioq_init(struct Qdisc *sch, struct nlattr *opt)
+{
+	int prio;
+	struct gkprioq_sched_data *q = qdisc_priv(sch);
+	struct Qdisc *pfifo_q;
+
+	if (opt == NULL) 
+		 q->max_limit = qdisc_dev(sch)->tx_queue_len; 
+	else {
+		struct tc_gkprioq_qopt *ctl = nla_data(opt);
+
+		if(nla_len(opt) < sizeof(*ctl))
+			return -EINVAL;
+
+		q->max_limit = ctl->limit;
+	}
+	
+	q->init_limit = 1;
+
+	for(prio = 0; prio <= 1; prio++) {
+		pfifo_q = fifo_create_dflt(sch, &pfifo_qdisc_ops, 
+					q->max_limit/(MAX_PRIORITY + 1));
+		if (IS_ERR(pfifo_q)) 
+			return PTR_ERR(pfifo_q);
+		q->prio_array[prio] = pfifo_q;
+	}
+
+	for(prio = 2; prio < MAX_PRIORITY + 1; prio++) {
+		pfifo_q = fifo_create_dflt(sch, &pfifo_qdisc_ops, 
+						q->init_limit);
+		if (IS_ERR(pfifo_q)) 
+			return PTR_ERR(pfifo_q);
+		q->prio_array[prio] = pfifo_q;
+	}
```

Comment:
Suppose when a high priority packet in ingressed but can not be accomodated by GK, just as it is, because GK is currently fully occupied with lower priority packets. Here, GK should drop the lowest priority packet and then enqueue the current ingress packet. This is essentially  what 'borrowing' does in this kernel module. And all the IF blocks are to handle the many cases of how this (dropping lowest priority packet and enqueueing the ingress higher priority packet)  should happen (limiting cases, special cases, etc)